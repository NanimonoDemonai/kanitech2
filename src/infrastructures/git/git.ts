import { simpleGit } from "simple-git";

const git = simpleGit();
git.init();

interface RawHistory {
  user: string;
  hash: string;
  message: string;
  date: string;
}

export interface History {
  user: string;
  hash: string;
  message: string;
  date: Date;
}

interface Param {
  file: string;
  hash: string;
}

/**
 * 記事のpidのgit historyを出す
 * @param file 記事のpid
 */
export const getHistory = async (file: string): Promise<readonly History[]> => {
  const log = await git.log<RawHistory>({
    file,
    format: {
      user: "%an",
      hash: "%H",
      message: "%s",
      date: "%ad",
    },
  });
  return log.all.map((e) => ({ ...e, date: new Date(e.date) }));
};

export const showHistory = async ({
  hash,
  file,
}: Param): Promise<
  | {
      success: true;
      data: string;
    }
  | { success: false; error: Error }
> => {
  try {
    const data = await git.show(`${hash}:${file}`);
    return { data, success: true };
  } catch (error) {
    if (error instanceof Error) {
      return { error, success: false };
    }
    return { error: new Error(`${hash}, ${file}`), success: false };
  }
};
