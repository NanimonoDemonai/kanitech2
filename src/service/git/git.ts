import { simpleGit } from "simple-git";

const git = simpleGit();
git.init();

interface RawHistory {
  user: string;
  hash: string;
  message: string;
  date: string;
}

export interface History {
  user: string;
  hash: string;
  message: string;
  date: Date;
}

interface Param {
  entry: string;
  hash: string;
}

export class Git {
  constructor(private path: string) {}

  /**
   * 記事のpidのgit historyを出す
   * @param entry 記事のpid
   */
  public getHistory = async (entry: string): Promise<readonly History[]> => {
    const log = await git.log<RawHistory>({
      file: `${this.path}/${entry}.mdx`,
      format: {
        user: "%an",
        hash: "%H",
        message: "%s",
        date: "%ad",
      },
    });
    return log.all.map((e) => ({ ...e, date: new Date(e.date) }));
  };

  public showHistory = async ({ hash, entry }: Param): Promise<string> =>
    git.show(`${hash}:${this.path}/${entry}.mdx`);
}
