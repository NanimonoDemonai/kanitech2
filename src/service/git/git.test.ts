import { getHistory, showHistory } from "src/service/git/git";

describe("git", () => {
  it("should get log", async () => {
    expect(await getHistory("how_to_mdx_bundling")).toMatchSnapshot();
  });
  it("should show History", async () => {
    expect(
      await showHistory({
        hash: "b1426f894423fff7b9cf679b566a1553508c15cd",
        entry: "how_to_mdx_bundling",
      })
    ).toMatchSnapshot();
  });
});
