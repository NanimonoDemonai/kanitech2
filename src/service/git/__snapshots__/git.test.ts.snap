// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`git should get log 1`] = `
[
  {
    "date": 2022-11-27T13:24:05.000Z,
    "hash": "fe4edc90e683cf9452d2ca6c9a330be19a661714",
    "message": "globするサービスを作る",
    "user": "nanimono_demonai",
  },
  {
    "date": 2022-11-20T08:50:18.000Z,
    "hash": "b1426f894423fff7b9cf679b566a1553508c15cd",
    "message": "コミットする",
    "user": "nanimono_demonai",
  },
  {
    "date": 2022-11-19T07:10:19.000Z,
    "hash": "eb50028da73216e8a46927c92f1272fd672aa960",
    "message": "pidでmdxをバンドルする",
    "user": "nanimono_demonai",
  },
  {
    "date": 2022-11-16T13:03:14.000Z,
    "hash": "2b960b9fb1ae0b86b61d0efed56f8c484a949740",
    "message": "mdxをコンパイルすることができるところまで",
    "user": "nanimono_demonai",
  },
]
`;

exports[`git should not show History 1`] = `
{
  "error": [Error: fatal: invalid object name 'hogehogehoge'.
],
  "success": false,
}
`;

exports[`git should show History 1`] = `
{
  "data": "---
title: "どのようにMDXをバンドルしているか"
tags: [mdx, nextjs, 解説記事]
---

# テスト1

## Sanguine ambae

Lorem markdownum ipsa ingemit messes te moenia meas neque, se nec. Obvertere
iuvenes faciemque. [Somnus](http://hoc.io/geminos.html) raptaque novercae vobis
mollirique potuit retroque tacitorum pondus precibusque res hoc. Adfixa velat
sumpsit traherent removerat quater ara populos. Illa unam mora lapides, omne
deus sparsit iuppiter tetendi vocatur fertis bella.

Per quae patriorum talibus omnes. Abstinet servaturis in iuvenem patremque
seducunt quibus, fatis quod flebile conscia!

## Fundae cultique conducat propulit

Vidisse in nomenque armis. Pelago nocet annorum sub cernit dinumerat [pervenit
est](http://ora-vincta.net/) antri imi magna, stabula et, semine: in: precor.
Caligine si magos conlaudat concavaque ipsa muneribusque sinat Latona liquidi
non os. Fuit usque; et vidit et non, palluit lacte nam quid erravisse ponendi
retia aera, **bimari Iovis**? Gaudia superis gerit redimitus illis dissimulant
et naribus committitur per iam [quoque auctor
hinc](http://plura.org/aperto.php).

> Ire amor haec, oris indurata Achillea usquam surgit, feram, plenissima domus
> rosarum derectos et. Mixta tum ignaroque secreta difficilis cornua hostem. In
> dis tunc est saecula, pollentibus tura caesariem fontesque.

\\\`\\\`\\\`javascript
alert("hello");
\\\`\\\`\\\`

## Omnia ingeniis utque

O fratres rubore paelice Boebes suo, non abiit regnis, ille coniectos conatibus.
Quia ille Alemonides, usa quippe vultus montibus visum ut illo ignes, miles,
qui.

## Praeferri et oscula medio tempus ferrugine litora

Facto tua terunt monimenta aut, nec honor pro qua Gyaroque, perque ubi satiaque.
Fulserunt sibila sinu concedimus, probavit sitim. Aut aspice paludes, maturus
centum damno. Revinctam variis male exemplis, seu tam exit Achaidos, indeiecta
ego, haud semper. Me raptam fronte cum hactenus aram latentis capiat pendet
sibi.

Ora ingens, dedisse nullaque movi opem furibunda tecum, ingenti? Rutuli Iuno
illa, Glaucus. Sinebat ala atque dis magna, ac in praecordia ignesque signa in
haec. Haut pro, cum ludit corpus; cum attollit
[navigat](http://cancer.org/melumina.aspx) occiduae ferinae certo _superi_, aut
ter!

\`",
  "success": true,
}
`;

exports[`git should show History 2`] = `
{
  "error": [Error: fatal: invalid object name 'hogehogehoge'.
],
  "success": false,
}
`;
